### PR TITLE
[Tree unique] Referencing Id Analysis

### PR DESCRIPTION
**Pull request by kirstenmg**
Creates relations `ExprHasRefId`, `ListExprHasRefId` to find the referencing id of an expression (the id of the context containing the expression).

### CODE REVIEW

**Comment by kirstenmg** [comment]
After merging with main, the following tests fail because the ref ids don't match:
`    simple::test_simple_bool_rules
    switch_rewrites::switch_pull_in_below
    switch_rewrites::test_constant_condition`
I will figure this out then re-open the PR.

File: tree_unique_args/src/schedule.egg [hunk 1]
```diff
@@ -3,6 +3,7 @@
 (run-schedule
   (repeat 5
     (saturate always-run)
+    (saturate error-checking) ; In the future, this will be "release mode" only
```
Comment (oflatt) [inline tree_unique_args/src/schedule.egg:6]: did you mean "debug mode" only?
I think it's fine to run these in release unless we think they are slow

File: tree_unique_args/src/id_analysis.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+(rule ((ListExprHasRefId x id1)
+       (ListExprHasRefId x id2)
+       (!= id1 id2))
+      ((panic \"Ref ids don't match\"))
+      :ruleset error-checking)
+    "
+    .to_string()];
+
+    ESort::iter()
+        .map(|sort| "(relation *HasRefId (* IdSort))".replace('*', sort.name()))
```
Comment (oflatt) [inline tree_unique_args/src/id_analysis.rs:48]: Here you map over the sorts, but above you implement for `Expr` and `ListExpr` manually.
Instead you could add your error checking rule in this templated string

**Review by oflatt — CHANGES_REQUESTED** [review]
Really great job on implementing this, testing, and fixing other tests.
Had a couple small comments

File: tree_unique_args/src/schedule.egg (hunk 1 above)
Comment (kirstenmg) [reply tree_unique_args/src/schedule.egg:6]: Yes, pretty sure this should be debug mode only (if other error-checking rules end up slow)

**Review by oflatt — APPROVED** [review]
Great work! Before we merge, could you add a test that it catches a couple bad cases?
You could add a couple
```
#[test]
#[should_panic]
```

**Pull request merged**